proposals/0000-reconsider-typealias-in-protocol.md

```diff
@@ -0,0 +1,138 @@
+has been a workaround for same-type constraints in protocol declarations, introduced together with [SE-0142](https://github.com/apple/swift-evolution/blob/master/proposals/0142-associated-types-constraints.md) in **Swift 4**.
+
+Here, the declaration of a `typealias` that shares its name with an `associatedtype` is tantamount to a same-type constraint
+that is syntactically and semantically inconsistent with the language. This feature violates the rule that declarations
+in a protocol extension should never conflict with its requirements (the `typealias` does not act as a default value) and
+introduces counterproductive situations that often have no practical value and are hard to judge and assess. 
+
+```swift
+protocol P {
+    associatedtype E
+    func foo(_ arg: E)
+}
+protocol P1: P {...} 
+
+class Foo: P {
+    typealias E = String
+
+    func foo(_ arg: E) {}
+}
+// This will break Foo and any other conformances to P and its descendants. 
+extension P { typealias E = Bool } 
+```
+A further contradiction is seen with the convention of leaving out any implementation details within a protocol body:
+those are ought to be described in an extension or upon conformance.
+
+### Type aliases in extensions
+Currently a `typealias` in a protocol extension that conflicts with an `associatedtype` is treated as a same-type constraint on the associatedtype **if** there are no appliances of the type name **among the requirements**. Example:
+
+```swift
+protocol P {associatedtype A}
+extension P {typealias A = Int} // same-type constraint on A.
+```
+
+If there **is** an appliance of `A`, an error occurs:
+
+```swift
+protocol P {
+  associatedtype A
+  func foo() -> A // 'A' is ambiguous for type lookup in this context
+}
+extension P {typealias A = Int}
+```
+
+This is inconsistent with the behavior of any other implementation details declared in protocol extensions, which become defaults. The fact that an appliance generates an error is either a bug or proof that the rules aren't solid and conflict with the language on the implementation level.
+
+## Proposed solution
+
+**Swift 4.1** began preparations to eliminate the feature by adding certain warnings and conditionally disallowing such type aliases in protocol extensions by raising a `type look up ambiguity` error towards appliances of the conflicting type name, if any.
+
+#### `#1`
+```swift
+protocol P {associatedtype A}
+protocol P1: P {typealias A = Int}
+// warning: typealias overriding associated type 'R' from protocol 'P' is better expressed as same-type constraint on the protocol
+```
+#### `#2`
+```swift
+protocol P {typealias A = Int}
+protocol P1: P {associatedtype A}
+// warning: associated type 'R' is redundant with type 'R' declared in inherited protocol 'P'
+```
+#### `#3`
+```swift
+protocol P {
+  associatedtype A
+  func foo() -> A // error: 'A' is ambiguous for type lookup in this context
```

Comment:
This is just a bug. I think what you really want to do is not ban protocol typealiases altogether, but change the rules for conformance checking to not consider typealiases in unconstrained protocol extensions as candidates for the associated type at all.